### DISCUSSION TITLE
5.9.10+swanctl+EAP client+xfrm: no route added ? no 220 table created

### DISCUSSION THREAD

**Discussion by frenchy59** [strongswan/strongswan#2546] [Q&A]
Hi,

I already have a remote IKEv2 VPN server working. (I can connect from IOS device via EAP_MSCHAPV2 method)
I am trying to setup strongswan-5.9.10-1.el9.x86_64 client on a freshly built CentOS 9 kernel 5.14.0-529.el9.x86_64
I am using the latest swanctl.conf (not the deprecated ipsec.conf method)

I was able to establish the connection succesfully,
/usr/sbin/swanctl --initiate --child home

but for some reason I cannot ping anything through the tunnel, anything I ping goes through my local network.
I also don't see any xfrm interface related created when running 'ip addr'
'ip route' don't show any additonnal route through the tunnel, so I am guessing I need to add one, but I would have thought this would be automatically added when I establish the connection ?
'ip xfrm policy' & 'ip xfrm state' are working, I can see the 2 flows:
src <vpn_client> dest <vpn_server>
src <vpn_server> dest <vpn_client>
both with if_id 0x1

'ip route list table 220' is not giving me anything, I am not sure if this table should be created automatically when I establish the connection ?
Error: ipv4: FIB table does not exist.
Dump terminated

Here is my config :
vi /etc/strongswan/swanctl/swanctl.conf :
connections {

   home {
      local_addrs=%any
      remote_addrs=jon.freeboxos.fr
      local {
         auth = eap
         eap_id = jon
      }
      remote {
         auth = pubkey
         id = jon.freeboxos.fr
      }
      children {
         home {
            local_ts = 0.0.0.0/0
            remote_ts = 0.0.0.0/0
            esp_proposals = aes256-sha256-modp2048
            if_id_in=%unique
            if_id_out=%unique
            mark_in=%unique
            mark_out=%unique
         }
      }
      version = 2
      proposals = aes256-sha256-modp2048
   }
}

secrets {

   eap-jon {
      id = jon
      secret = mypassword
   }
}

I had a read through https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html
since 5.9.10, my understanding is that the route should be added automatically if install_routes_xfrmi = yes

cat /etc/strongswan/strongswan.d/charon/kernel-netlink.conf | grep -Ev "#|^$" :

kernel-netlink {
install_routes_xfrmi = yes
    load = yes
    spdh_thresh {
        ipv4 {
        }
        ipv6 {
        }
    }
}

not sure what I am missing here, but the route is not getting created  and I don't see any error/warning in 'journalctl -u strongswan' related to route so I don't know if it even tries to add it ...
any thoughts ?

**Comment by tobiasbrunner** [comment]
> if_id_in=%unique
> if_id_out=%unique
> ...
> I had a read through https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html
> since 5.9.10, my understanding is that the route should be added automatically if install_routes_xfrmi = yes

Only if an interface with matching interface ID is found when the SAs and policies are installed. However, with `%unique` that will not be the case as every CHILD_SA gets a new unique ID.

**Comment by frenchy59** [comment]
Thanks Tobias, appreciate the answer.
I have tested the same with the following in /etc/strongswan/swanctl/swanctl.conf :
if_id_in=3
if_id_out=3

and also, I have added the fwmark = 0x2  in  /etc/strongswan/strongswan.d/charon/socket-default.conf

I also tried to add/remove the following in /etc/strongswan/swanctl/swanctl.conf but not to avail :
set_mark_out=0x1

still the same issue, the `ip -s xfrm state`is now showing id 0x3 rather than 0x1 but still the same the routes are missing, I don't see any table 220 neither.

please advise.

**Reply by tobiasbrunner** [reply]
Did you create an XFRM interface with that ID?

**Comment by frenchy59** [comment]
initially I was not too sure if I had to manually create this interface or not ? also whether this interface should be created before or after I start strongswan service and establish the child "home" connection ?

I did try to create it anyway, for investigation purpose, but I am not too sure if I do it properly:
ip link add ipsec0 type xfrm if_id 3
ip link set ipsec0 up

so far so good, but then, I am not sure if I should set an IP on it, which IP I should set
and also whether i need to create the route afterwards?

**Reply by tobiasbrunner** [reply]
> initially I was not too sure if I had to manually create this interface or not ? also whether this interface should be created before or after I start strongswan service and establish the child "home" connection ?

Depends on your use case. But if you want to use the `install_routes_xfrmi` feature, you have to create it before establishing the connection.

> so far so good, but then, I am not sure if I should set an IP on it, which IP I should set

Also depends on the use case. If you use virtual IPs, you could e.g. let them get installed on it (via `install_virtual_ip_on` option). But on Linux it doesn't really matter on which interface an IP is installed.

> and also whether i need to create the route afterwards?

Not if you use the `install_routes_xfrmi` feature.

**Comment by frenchy59** [comment]
I thought "install_virtual_ip = yes" was the default but I have now set it explicitly on /etc/strongswan/strongswan.d/charon.conf
then ran following commands:
`# ip link add ipsec0 type xfrm if_id 3`
`# ip link set ipsec0 up`
`# systemctl restart strongswan`
`# /usr/sbin/swanctl --initiate --child home`
...
[IKE] peer supports MOBIKE
[IKE] IKE_SA home[1] established between 192.168.1.59[192.168.1.59]...89.165.XX.XX[jon.freeboxos.fr]
[IKE] scheduling rekeying in 13337s
[IKE] maximum IKE_SA lifetime 14777s
[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
[IKE] CHILD_SA home{1} established with SPIs cdc1eda5_i c0c25902_o and TS 95.240.X.X/32 === 0.0.0.0/0
initiate completed successfully

`# /usr/sbin/swanctl --list-sas`
plugin 'sqlite': failed to load - sqlite_plugin_create not found and no plugin file available
home: #1, ESTABLISHED, IKEv2, 7881002472923196_i* 82ee3f08528214d9_r
  local  '192.168.1.59' @ 192.168.1.59[4500]
  remote 'jon.freeboxos.fr' @ 89.165.XX.XX[4500]
  AES_CBC-256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
  established 106s ago, rekeying in 13231s
  home: #1, reqid 1, INSTALLED, TUNNEL-in-UDP, ESP:AES_CBC-256/HMAC_SHA2_256_128
    installed 107s ago, rekeying in 3303s, expires in 3854s
    in  cdc1eda5 (-|0x00000003),      0 bytes,     0 packets
    out c0c25902 (-|0x00000003),      0 bytes,     0 packets
    local  95.240.X.X/32
    remote 0.0.0.0/0


`# ip a`
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: ens33: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0c:29:77:0e:7d brd ff:ff:ff:ff:ff:ff
    altname enp2s1
    inet 192.168.1.59/24 brd 192.168.1.255 scope global noprefixroute ens33
       valid_lft forever preferred_lft forever
    inet6 fe80::20c:29ff:fe77:e7d/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
5: ipsec0@NONE: <NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none
    inet6 fe80::2f11:213b:7961:ba4e/64 scope link stable-privacy
       valid_lft forever preferred_lft forever


--> The virtual IP has not been set ... and the routes neither.


`# ip route`
default via 192.168.1.1 dev ens33 proto dhcp src 192.168.1.59 metric 100
192.168.1.0/24 dev ens33 proto kernel scope link src 192.168.1.59 metric 100

`# ip route list table 220`
Error: ipv4: FIB table does not exist.
Dump terminated


`# ip xfrm state`
src 192.168.1.59 dst 89.165.XX.XX
        proto esp spi 0xc0c25902 reqid 1 mode tunnel
        replay-window 0 flag af-unspec
        output-mark 0x1/0xffffffff
        auth-trunc hmac(sha256) 0x0f32363074e288c15517929aff0b0d1bb49c7c0fb03c6900591c435cee9670e9 128
        enc cbc(aes) 0x7d3b9afaab0644bdf9d23c85baf7292cc530564dd483bd340d82f3f76045673e
        encap type espinudp sport 4500 dport 4500 addr 0.0.0.0
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        if_id 0x3
src 89.165.XX.XX dst 192.168.1.59
        proto esp spi 0xcdc1eda5 reqid 1 mode tunnel
        replay-window 32 flag af-unspec
        auth-trunc hmac(sha256) 0x42c7ca1a46987a00cb47296dc9ab01bd7cc9fc3b04d8decc5cdb7ea7df21c2b4 128
        enc cbc(aes) 0x200f1fa629120a543db1e1ecb0fd3f5830eb9ac55fe32594820b59436c8bea26
        encap type espinudp sport 4500 dport 4500 addr 0.0.0.0
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        if_id 0x3

`# ping google.com`
PING google.com (142.250.200.14) 56(84) bytes of data.
64 bytes from lhr48s29-in-f14.1e100.net (142.250.200.14): icmp_seq=1 ttl=119 time=8.74 ms
64 bytes from lhr48s29-in-f14.1e100.net (142.250.200.14): icmp_seq=2 ttl=119 time=9.64 ms
64 bytes from lhr48s29-in-f14.1e100.net (142.250.200.14): icmp_seq=3 ttl=119 time=9.50 ms
64 bytes from lhr48s29-in-f14.1e100.net (142.250.200.14): icmp_seq=4 ttl=119 time=9.57 ms
64 bytes from lhr48s29-in-f14.1e100.net (142.250.200.14): icmp_seq=5 ttl=119 time=9.64 ms
^C
--- google.com ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4007ms
rtt min/avg/max/mdev = 8.737/9.416/9.642/0.343 ms


--> but the traffic still goes through my local NAT/internet, and not through the VPN, you can see there is no traffic detected, 0 packets :

`# /usr/sbin/swanctl --list-sas`
plugin 'sqlite': failed to load - sqlite_plugin_create not found and no plugin file available
home: #1, ESTABLISHED, IKEv2, 7881002472923196_i* 82ee3f08528214d9_r
  local  '192.168.1.59' @ 192.168.1.59[4500]
  remote 'jon.freeboxos.fr' @ 89.165.XX.XX[4500]
  AES_CBC-256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
  established 470s ago, rekeying in 12867s
  home: #1, reqid 1, INSTALLED, TUNNEL-in-UDP, ESP:AES_CBC-256/HMAC_SHA2_256_128
    installed 471s ago, rekeying in 2939s, expires in 3490s
    in  cdc1eda5 (-|0x00000003),      0 bytes,     0 packets
    out c0c25902 (-|0x00000003),      0 bytes,     0 packets
    local  95.240.X.X/32
    remote 0.0.0.0/0


--> Also, my public IP is still my local one, not the VPN one :

`# curl ifconfig.me`
95.240.X.X

**Reply by tobiasbrunner** [reply]
> I thought "install_virtual_ip = yes" was the default

Not `install_virtual_ip`, `install_virtual_ip_on`, which takes the name of an interface to which virtual IPs are added.

> --> The virtual IP has not been set

That's because you have not configured the negotiation of [virtual IPs](https://docs.strongswan.org/docs/latest/features/vip.html#_initiator_configuration).

> and the routes neither.

Check the log if you are interested in details (use the log levels in #196). But the problem is that the IP address `95.240.X.X/32` is obviously not a local IP address that could be used as source address but your public address as seen by the server. So this won't work without the use of virtual IPs.
